lib/base/json.cpp

```diff
@@ -1,23 +1,228 @@
 /* Icinga 2 | (c) 2012 Icinga GmbH | GPLv2+ */
 
 #include "base/json.hpp"
-#include "base/debug.hpp"
-#include "base/namespace.hpp"
-#include "base/dictionary.hpp"
 #include "base/array.hpp"
-#include "base/objectlock.hpp"
-#include "base/convert.hpp"
-#include "base/utility.hpp"
-#include <bitset>
-#include <boost/exception_ptr.hpp>
-#include <cstdint>
-#include <json.hpp>
-#include <stack>
-#include <utility>
-#include <vector>
+#include "base/generator.hpp"
 
 using namespace icinga;
 
+constexpr uint8_t JsonEncoder::m_IndentSize;
+
+JsonEncoder::JsonEncoder(std::string& output, bool prettify)
+	: JsonEncoder{nlohmann::detail::output_adapter<char>(output), prettify}
+{
+}
+
+JsonEncoder::JsonEncoder(std::basic_ostream<char>& stream, bool prettify)
+	: JsonEncoder{nlohmann::detail::output_adapter<char>(stream), prettify}
+{
+}
+
+JsonEncoder::JsonEncoder(nlohmann::detail::output_adapter_t<char> stream, bool prettify)
+	: m_Pretty(prettify), m_Indent{0}, m_IndentStr(32, ' '), m_Writer(std::move(stream))
+{
+}
+
+/**
+ * Emits JSON token(s) based on the provided control flag.
+ *
+ * This function handles various control flags that dictate how the JSON structure should be encoded,
+ * and it manages the context stack to ensure proper nesting of JSON objects and arrays.
+ *
+ * @param flag The control flag indicating the types of JSON structure to encode.
+ */
+void JsonEncoder::EncodeImpl(Ctrls flag)
+{
+	switch (flag) {
+		case BeginObject:
+			m_CtxStack.push(Ctx{CtxType::Object, 0});
+			Indent();
+			Write("{");
+			return;
+		case EndObject:
+			// Ensure that we are indeed in an object context before closing it.
+			ASSERT(!m_CtxStack.empty() && m_CtxStack.top().type == CtxType::Object);
+			// UnIndent will emit a newline and indentation if the Object has elements,
+			// so we must call it before popping the context stack.
+			UnIndent();
+			m_CtxStack.pop();
+			Write("}");
+			return;
+		case BeginArray:
+			m_CtxStack.push(Ctx{CtxType::Array, 0});
+			Indent();
+			Write("[");
+			return;
+		case EndArray:
+			// Ensure that we are indeed in an array context before closing it.
+			ASSERT(!m_CtxStack.empty() && m_CtxStack.top().type == CtxType::Array);
+			// UnIndent will emit a newline and indentation if the array has elements,
+			// so we must call it before popping the context stack.
+			UnIndent();
+			m_CtxStack.pop();
+			Write("]");
+			return;
+		case Indentation:
+			m_Writer->write_characters(m_IndentStr.c_str(), m_Indent);
+			return;
+		case Comma:
+			Write(m_Pretty ? ",\n" : ",");
```

Comment:
```suggestion
			Write(",");
```

... if you don't write \\n in advance (just like with the array).